overlord/fdestate/fdestate.go

```diff
@@ -0,0 +1,322 @@
+func (h hashAlg) MarshalJSON() ([]byte, error) {
+	switch crypto.Hash(h) {
+	case crypto.SHA256:
+		return json.Marshal("sha256")
+	default:
+		return nil, fmt.Errorf("Unknown algorithm %v", h)
+	}
+}
+
+// KeyDigest stores a HMAC(key, salt) of a key
+type KeyDigest struct {
+	// Algorithm is the algorithm for
+	Algorithm hashAlg `json:"alg"`
+	// Salt is the salt for the HMAC digest
+	Salt []byte `json:"salt"`
+	// Digest is the result of `HMAC(key, salt)`
+	Digest []byte `json:"digest"`
+}
+
+const defaultHashAlg = crypto.SHA256
+
+func getPrimaryKeyDigest(devicePath string) (KeyDigest, error) {
+	salt, digest, err := secbootGetPrimaryKeyHash(devicePath, crypto.Hash(defaultHashAlg))
+	if err != nil {
+		return KeyDigest{}, err
+	}
+
+	return KeyDigest{
+		Algorithm: hashAlg(defaultHashAlg),
+		Salt:      salt,
+		Digest:    digest,
+	}, nil
+}
+
+func (kd *KeyDigest) verifyPrimaryKeyDigest(devicePath string) (bool, error) {
+	return secbootVerifyPrimaryKeyHash(devicePath, crypto.Hash(kd.Algorithm), kd.Salt, kd.Digest)
+}
+
+// PrimaryKeyInfo provides information about a primary key that is used to manage key slots
+type PrimaryKeyInfo struct {
+	Digest KeyDigest `json:"digest"`
+}
+
+// FdeState is the root persistent FDE state
+type FdeState struct {
+	// PrimaryKeys are the keys on the system. Key with ID 0 is
+	// reserved for snapd and is populated on first boot. Other
+	// IDs are for externally managed keys.
+	PrimaryKeys map[int]PrimaryKeyInfo `json:"primary-keys"`
+
+	// KeyslotRoles are all keyslot roles indexed by the role name
+	KeyslotRoles map[string]KeyslotRoleInfo `json:"keyslot-roles"`
+}
+
+const fdeStateKey = "fde"
+
+func initializeState(st *state.State) error {
+	var s FdeState
+	err := st.Get(fdeStateKey, &s)
+	if err == nil {
+		// TODO: Do we need to do something in recover?
+		return nil
+	}
+
+	if !errors.Is(err, state.ErrNoState) {
+		return err
+	}
+
+	dataUUID, dataErr := disksDMCryptUUIDFromMountPoint(dirs.GlobalRootDir)
+	saveUUID, saveErr := disksDMCryptUUIDFromMountPoint(dirs.SnapSaveDir)
+	if errors.Is(saveErr, &disks.ErrMountPointNotFound{}) {
+		// TODO: do we need to care about old cases where there is no save partition?
+		return nil
+	}
+
+	if errors.Is(dataErr, disks.ErrNoDmUUID) && errors.Is(saveErr, disks.ErrNoDmUUID) {
+		return nil
+	}
+	if dataErr != nil {
+		return dataErr
+	}
+	if saveErr != nil {
+		return saveErr
+	}
+
+	digest, err := getPrimaryKeyDigest(fmt.Sprintf("/dev/disk/by-uuid/%s", dataUUID))
+	if err != nil {
+		return err
+	}
+	sameDigest, err := digest.verifyPrimaryKeyDigest(fmt.Sprintf("/dev/disk/by-uuid/%s", saveUUID))
+	if err != nil {
+		return err
+	}
+	if !sameDigest {
+		return fmt.Errorf("Primary key for data and save partition are not the same")
+	}
+
+	s.PrimaryKeys = map[int]PrimaryKeyInfo{
+		0: {
+			Digest: digest,
+		},
+	}
+
+	// Note that Parameters will be updated on first update
+	s.KeyslotRoles = map[string]KeyslotRoleInfo{
+		"run+recover": {
```

Comment:
feels like this should be a const somewhere to rule out use of say `recover+run`

Reply:
The problem here is that we need that value both in `overlord/fdestate` and `overlord/fdestate/backend`. But `overlord/fdestate/backend` can not import `overlord/fdestate`. So it has to be in the backend. Unless we create new package. @pedronis what do you think?

Reply:
let's leave TODOs for now, I wonder which other places will need these string constants

Reply:
Done